File: LinkedList /delmid.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def delete_middle(head):
    if head is None or head.next is None:
        return None

    slow = head
    fast = head
    prev = None

    while fast and fast.next:
        fast = fast.next.next
        prev = slow
        slow = slow.next

    prev.next = slow.next

    return head

def print_list(head):
    temp = head
    while temp:
        print(temp.data, end=" -> ")
        temp = temp.next
    print("None")

File: LinkedList /delmid.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class DeleteMiddle {
    
    public static Node deleteMiddle(Node head) {
        
        if (head == null || head.next == null) {
            return null;
        }

        Node slow = head;
        Node fast = head.next; 
        Node prev = null;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            prev = slow;
            slow = slow.next;
        }

   
        if (prev != null) {
            prev.next = slow.next;
        } else {
            head = head.next; 
        }

        return head;
    }

    public static void printList(Node head) {
        Node curr = head;
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.next;
        }
        System.out.println();
    }
}
